File: server/services/activity/activity_watch_service.py
from typing import Dict, List, Optional, Tuple, Union
from dataclasses import dataclass
from datetime import datetime, timezone
import logging
import json

from aw_client import ActivityWatchClient
from aw_client.queries import canonicalEvents, DesktopQueryParams

# Configure logging to see error messages
logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


@dataclass
class AppUsagePattern:
    """Represents the usage pattern analysis for a time window."""
    dominant_app: str
    dominance_ratio: float
    app_count: int
    is_productive: bool
    sorted_apps: List[Tuple[str, float]]

@dataclass
class SessionWindowStates:
    """Represents state information extracted from session window bucket history."""
    sequence: List[str]
    last: str
    consecutive_prefocus: int
    consecutive_unproductive: int

@dataclass
class ActivityBucket:
    """Represents a single activity bucket from the session window."""
    startTime: str
    endTime: str
    category: str
    id: Optional[int] = None
    subtaskId: Optional[int] = None
    date: Optional[str] = None
    dominantApp: Optional[str] = None
    apps: Optional[Dict[str, float]] = None
    workSessionApp: Optional[str] = None
    createdAt: Optional[str] = None
    updatedAt: Optional[str] = None 

class Config:
    """Constants for thresholds and configuration."""
    IDLE_THRESHOLD_PERCENTAGE = 0.1  # Minimum 10% of window time must have activity to avoid idle classification
    WORK_SESSION_APP_THRESHOLD_PERCENTAGE = 0.05  # Minimum 5% of bucket time for meaningful app usage
    PREFOCUS_BUCKETS_REQUIRED = 2   # Consecutive prefocus buckets needed to achieve focus
    MAX_DISTRACTION_BUCKETS = 2     # Maximum distraction buckets before focus reset
    
    # Focus pattern threshold
    PRODUCTIVE_SWITCHING_THRESHOLD = 0.5   # Dominance ratio for productive classification
    
    # App switching distraction detection
    APP_SWITCHING_WINDOW_SIZE = 4   # Number of recent buckets to analyze for app switching
    APP_SWITCHING_THRESHOLD = 0.75  # If 75% or more buckets have different dominant apps, classify as distraction


class ActivityWatchService:
    def __init__(self, testing: bool = False):
        """Initialize the ActivityWatch service with configuration."""
        self.client = ActivityWatchClient("kairo", testing=testing)
        self.config = Config()
    
    @staticmethod
    def parse_datetime(dt_str: str) -> datetime:
        """Simple datetime parsing that handles ISO format strings and ensures timezone-aware output."""
        if not dt_str:
            raise ValueError("Empty datetime string")
        
        # Handle Z suffix by converting to +00:00 format for fromisoformat compatibility
        if dt_str.endswith('Z'):
            dt_str = dt_str.replace('Z', '+00:00')
        
        # Parse the ISO format string
        dt = datetime.fromisoformat(dt_str)
        
        # If somehow timezone-naive, assume UTC
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        
        return dt
    
    def analyze_window_activity(
        self, 
        start_time: datetime, 
        end_time: datetime, 
        session_window: Optional[List[ActivityBucket]] = None
    ) -> Tuple[str, str, Dict[str, float], str]:
        """Analyze activity in a specific time window with historical context."""
        try:
            logger.info(f"Starting analysis for {start_time} to {end_time}")
            
            # Step 1: Get Activity Watch data
            buckets = self.client.get_buckets()
            window_bucket = afk_bucket = None
            
            for bucket_id in buckets.keys():
                if 'window' in bucket_id:
                    window_bucket = bucket_id
                elif 'afk' in bucket_id:
                    afk_bucket = bucket_id
            
            if not window_bucket:
                raise Exception("No window bucket found - ActivityWatch may not be running")
                
            # Set up query parameters for desktop with minimal categorization
            params = DesktopQueryParams(
                bid_window=window_bucket,
                bid_afk=afk_bucket or window_bucket,
                classes=[] # no categorization
            )
            
            query = canonicalEvents(params)
            
            # Add custom RETURN to get just the events and avoid categorization issues
            query = f"""
                    {query}
                    RETURN = {{"events": events}};
                    """
            # Execute the query for the specified time period
            result = self.client.query(query, [(start_time, end_time)])
            
            logger.info(f"Query result: {result}")
            
            # Extract events from the result
            if not result or not result[0] or not result[0].get("events"):
                logger.error(f"Invalid result structure: result={result}")
                raise Exception("Canonical events failed: result is invalid")
                
            events = result[0]["events"]  # Get events from the custom return structure
            
            # Step 2: Process activity data into app usage patterns
            window_duration_seconds = (end_time - start_time).total_seconds()
            app_usage = self._calculate_app_usage(events)
            if self._is_idle_period(app_usage, window_duration_seconds):
                return 'idle', '', app_usage.get('apps', {}), ''
            
            # Step 3: Analyze app usage patterns for focus classification
            app_patterns = self._analyze_app_patterns(
                app_usage['apps'], 
                app_usage['total_activity']
            )
            
            # Step 4: Determine work session context first
            work_session_app = self._detect_work_session_app(session_window) if session_window else ''
            
            # Step 5: Apply state machine logic with historical context
            category = self._classify_activity_state(app_patterns, session_window or [], work_session_app)
            
            return category, app_patterns.dominant_app, app_usage['apps'], work_session_app
                
        except Exception as e:
            logger.error(f"Error analyzing window activity: {e}")
            return 'idle', '', {}, ''
    
    def _calculate_app_usage(self, events) -> Dict[str, Union[Dict[str, float], float]]:
        """Calculate app usage statistics from ActivityWatch events."""
        apps = {}
        total_activity = 0
        
        for event in events:
            app_name = event.get('data', {}).get('app', '').lower()
            duration_seconds = event.get('duration', 0)
            
            total_activity += duration_seconds
            
            if app_name:
                # Store duration in seconds to match total_activity units
                apps[app_name] = apps.get(app_name, 0) + duration_seconds
        
        return {
            'apps': apps,
            'total_activity': total_activity
        }
    
    def _is_idle_period(self, app_usage: Dict, window_duration_seconds: float) -> bool:
        """Check if the time period should be classified as idle based on activity percentage."""
        if not app_usage['apps'] or window_duration_seconds <= 0:
            return True
            
        activity_percentage = app_usage['total_activity'] / window_duration_seconds
        return activity_percentage < self.config.IDLE_THRESHOLD_PERCENTAGE
    
    def _analyze_app_patterns(self, apps: Dict[str, float], total_activity: float) -> AppUsagePattern:
        """Analyze app usage patterns to determine focus characteristics."""
        # Handle empty apps or zero total activity
        if not apps or total_activity <= 0:
            return AppUsagePattern(
                dominant_app='unknown',
                dominance_ratio=0.0,
                app_count=0,
                is_productive=False,
                sorted_apps=[]
            )
        
        # Calculate dominance metrics
        sorted_apps = sorted(apps.items(), key=lambda x: x[1], reverse=True)
        dominant_app, dominant_time_seconds = sorted_apps[0]
        dominance_ratio = dominant_time_seconds / total_activity
        
        # Determine if activity is productive
        is_productive = dominance_ratio >= self.config.PRODUCTIVE_SWITCHING_THRESHOLD
        
        return AppUsagePattern(
            dominant_app=dominant_app,
            dominance_ratio=dominance_ratio,
            app_count=len(apps),
            is_productive=is_productive,
            sorted_apps=sorted_apps
        )
    
    def _classify_activity_state(
        self, 
        app_patterns: AppUsagePattern, 
        session_window: List[ActivityBucket],
        work_session_app: str
    ) -> str:
        """Classify the current activity state using app patterns and historical context."""
        # Determine if current activity is productive
        is_productive = app_patterns.is_productive
        
        # Check for app switching distraction patterns in session window
        has_app_switching_distraction = self._detect_app_switching_distraction(session_window, work_session_app, app_patterns.dominant_app)
        
        # Handle classification without historical context
        if not session_window:
            # Without history, respect the pattern quality
            if is_productive and not has_app_switching_distraction:
                return 'prefocus'  # Start building toward focus
            else:
                return 'distraction'
        
        # Analyze historical context
        recent_states = self._extract_recent_states(session_window)
        
        # Apply state machine logic with full context
        if is_productive and not has_app_switching_distraction:
            return self._handle_productive_state(recent_states)
        else:
            return 'distraction'
    
    def _extract_recent_states(self, session_window: List[ActivityBucket]) -> SessionWindowStates:
        """Extract state information from session window bucket history."""
        state_sequence = [bucket.category for bucket in session_window]
        last_state = state_sequence[-1] if state_sequence else 'unknown'
        
        return SessionWindowStates(
            sequence=state_sequence,
            last=last_state,
            consecutive_prefocus=self._count_consecutive_from_end(state_sequence, 'prefocus'),
            consecutive_unproductive=self._count_consecutive_from_end(state_sequence, ['distraction', 'idle'])
        )

    def _detect_app_switching_distraction(self, session_window: List[ActivityBucket], work_session_app: str, current_dominant_app: str) -> bool:
        """Detect if there's excessive app switching indicating distraction."""
        if not session_window or len(session_window) < 2:
            return False
        
        # Start from the newest buckets and work backwards
        # Only consider buckets that are in focused states (focus/prefocus)
        focused_buckets = []
        for bucket in reversed(session_window):
            if bucket.category in ['focus', 'prefocus'] and bucket.dominantApp:
                # Stop if we encounter a different work session app (different work context)
                if bucket.workSessionApp and bucket.workSessionApp != work_session_app:
                    break
                focused_buckets.append(bucket)
                # Stop when we have enough buckets to analyze
                if len(focused_buckets) >= self.config.APP_SWITCHING_WINDOW_SIZE:
                    break
        
        # Extract dominant apps from focused buckets (newest first)
        dominant_apps = [bucket.dominantApp for bucket in focused_buckets]
        
        # Add current dominant app if valid (represents the current activity)
        if current_dominant_app and current_dominant_app != 'unknown':
            dominant_apps.append(current_dominant_app)
        
        # Need at least 2 apps to detect switching
        if len(dominant_apps) < 2:
            return False
        
        # Count unique apps in the recent focused window
        unique_apps = set(dominant_apps)
        unique_app_ratio = len(unique_apps) / len(dominant_apps)
        
        # If the ratio of unique apps is high, it indicates frequent switching
        is_switching = unique_app_ratio >= self.config.APP_SWITCHING_THRESHOLD
        
        logger.debug(f"App switching analysis (focused only): current_dominant_app={current_dominant_app}, work_session_app={work_session_app}, apps={dominant_apps}, unique_ratio={unique_app_ratio:.2f}, threshold={self.config.APP_SWITCHING_THRESHOLD}, is_switching={is_switching}")
        
        return is_switching
    
    def _handle_productive_state(
        self, 
        recent_states: SessionWindowStates, 
    ) -> str:
        """Handle classification when current activity is productive (focused patterns)."""
        
        # If previous state was focus, maintain focus (momentum)
        if recent_states.last == 'focus':
            return 'focus'
        elif recent_states.consecutive_prefocus >= self.config.PREFOCUS_BUCKETS_REQUIRED - 1:
            # Accumulated enough prefocus to achieve focus
            return 'focus'
        elif recent_states.consecutive_unproductive > 0 and recent_states.consecutive_unproductive <= self.config.MAX_DISTRACTION_BUCKETS and self._had_focus_before_distractions(recent_states.sequence):
            return 'focus'
        else:
            # Still building prefocus
            return 'prefocus'
    
    def _count_consecutive_from_end(self, states: List[str], target_state: Union[str, List[str]]) -> int:
        """Count consecutive occurrences of target_state(s) from the end of the list."""
        count = 0
        # Convert single string to list for uniform handling
        target_states = [target_state] if isinstance(target_state, str) else target_state
        
        for state in reversed(states):
            if state in target_states:
                count += 1
            else:
                break
        return count

    def _had_focus_before_distractions(self, states: List[str]) -> bool:
        """Check if there was a focus state directly before the current consecutive distractions/idle."""
        if not states:
            return False
        
        # Start from the end and skip consecutive distractions/idle
        distraction_states = {'distraction', 'idle'}
        focus_check_index = len(states)
        
        # Skip consecutive distractions/idle from the end
        for i in range(len(states) - 1, -1, -1):
            if states[i] in distraction_states:
                focus_check_index = i
            else:
                break
        
        # Check if the state directly before the distractions is focus
        if focus_check_index > 0:
            return states[focus_check_index - 1] == 'focus'
        
        return False

    def _detect_work_session_app(self, session_window: List[ActivityBucket]) -> str:
        """Detect the primary app being used in the current work session."""
        if not session_window:
            return ''
        
        app_total_time = {}
        
        for bucket in session_window:
            if bucket.category in ['focus', 'prefocus']:
                # Calculate bucket duration for percentage calculation using unified parsing
                bucket_start = self.parse_datetime(bucket.startTime)
                bucket_end = self.parse_datetime(bucket.endTime)
                bucket_duration_seconds = (bucket_end - bucket_start).total_seconds()
                
                # Get apps usage data from bucket
                apps_data = bucket.apps or {}
                if isinstance(apps_data, dict):
                    for app_name, usage_time in apps_data.items():
                        # Calculate usage percentage for this bucket
                        usage_percentage = usage_time / bucket_duration_seconds if bucket_duration_seconds > 0 else 0
                        # Only count apps with meaningful usage (avoid noise)
                        if usage_percentage >= self.config.WORK_SESSION_APP_THRESHOLD_PERCENTAGE:
                            app_total_time[app_name] = app_total_time.get(app_name, 0) + usage_time
        
        if not app_total_time:
            return ''
        
        # Return app with highest total time usage
        return max(app_total_time.keys(), key=lambda k: app_total_time[k])


def main(start_time_str: str, end_time_str: str, session_window: List[Dict] = None, testing: bool = False) -> None:
    try:
        # Parse timezone-aware datetime using unified utility
        start_time_tz = ActivityWatchService.parse_datetime(start_time_str)
        end_time_tz = ActivityWatchService.parse_datetime(end_time_str)
        

        # Initialize ActivityWatch service
        service = ActivityWatchService(testing=testing)
        
        # Convert session window and sort by startTime if provided
        if session_window:
            # Convert to ActivityBucket objects using dictionary unpacking
            session_window = [ActivityBucket(**bucket) for bucket in session_window]
            # Sort by startTime to ensure chronological order (oldest to newest)
            session_window.sort(key=lambda x: ActivityWatchService.parse_datetime(x.startTime))
        
        # Perform activity analysis
        category, dominant_app, apps, work_session_app = service.analyze_window_activity(
            start_time_tz, end_time_tz, session_window
        )
        
        # Create result using ActivityBucket dataclass with timezone-aware timestamps
        result = ActivityBucket(
            startTime=start_time_str,
            endTime=end_time_str,
            category=category,
            dominantApp=dominant_app if category != 'idle' else None,
            apps=apps,
            workSessionApp=work_session_app if work_session_app else None
        )
        
        # Convert to dict for JSON output
        print(json.dumps(result.__dict__, indent=2))
        
    except Exception as e:
        logger.error(f"Error in main analysis: {e}")
        # Return empty result on error to maintain API contract
        error_result = ActivityBucket(
            startTime=start_time_str if 'start_time_str' in locals() else '',
            endTime=end_time_str if 'end_time_str' in locals() else '',
            category='idle',
            dominantApp=None,
            apps={},
            workSessionApp=None
        )
        print(json.dumps(error_result.__dict__, indent=2))


if __name__ == '__main__':
    import argparse
    
    parser = argparse.ArgumentParser(description='Kairo ActivityWatch Processor')
    parser.add_argument('--start', required=True, help='Start time (ISO format)')
    parser.add_argument('--end', required=True, help='End time (ISO format)')
    parser.add_argument('--session-window', type=str, help='Session window data as JSON string')
    parser.add_argument('--testing', action='store_true', help='Use testing buckets')
    
    args = parser.parse_args()
    
    # Parse session window if provided
    session_window = None
    if args.session_window:
        try:
            session_window = json.loads(args.session_window)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse session window JSON: {e}")
            session_window = []
    
    main(args.start, args.end, session_window, args.testing)

File: server/services/activity/activityWatchService.ts
/**
 * ActivityWatch Service — TypeScript port of activity_watch_service.py
 *
 * Analyses a time window via the AW REST API, classifies it as
 * focus / prefocus / distraction / idle, and returns a result
 * that nodeActivityWatch.ts persists to the database.
 */

import { ActivityWatchClient, type AWEvent } from './activityWatchClient';
import type { ActivityBucket } from '../../../shared/schema';

// ─── Configuration ──────────────────────────────────────────────────────────────

const CONFIG = {
  /** Minimum percentage of window time with activity to avoid idle */
  IDLE_THRESHOLD_PERCENTAGE: 0.25,
  /** Minimum percentage of bucket time for meaningful app usage */
  WORK_SESSION_APP_THRESHOLD_PERCENTAGE: 0.05,
  /** Consecutive prefocus buckets needed to reach focus */
  PREFOCUS_BUCKETS_REQUIRED: 2,
  /** Maximum consecutive distraction buckets before focus resets */
  MAX_DISTRACTION_BUCKETS: 2,
  /** Dominance ratio threshold for productive classification */
  PRODUCTIVE_SWITCHING_THRESHOLD: 0.4,
  /** Recent focused-bucket window for switching analysis */
  APP_SWITCHING_WINDOW_SIZE: 4,
  /** Unique-app ratio that triggers distraction */
  APP_SWITCHING_THRESHOLD: 0.75,
  /** Max apps in a bucket to qualify for neutral (prefocus) instead of distraction */
  NEUTRAL_MAX_APP_COUNT: 3,
  /** Number of top apps per bucket to use for switching detection */
  SWITCHING_TOP_N_APPS: 2,
} as const;

// ─── Internal types ─────────────────────────────────────────────────────────────

interface AppUsagePattern {
  dominantApp: string;
  dominanceRatio: number;
  appCount: number;
  isProductive: boolean;
  sortedApps: [string, number][];
}

interface SessionWindowStates {
  sequence: string[];
  last: string;
  consecutivePrefocus: number;
  consecutiveUnproductive: number;
}

/** Valid activity classification categories. */
export type ActivityCategory = 'focus' | 'prefocus' | 'distraction' | 'idle';

/** The subset of ActivityBucket fields the service produces (no DB fields). */
export interface AnalysisResult {
  startTime: string;
  endTime: string;
  category: ActivityCategory;
  dominantApp: string | null;
  apps: Record<string, number>;
  workSessionApp: string | null;
}

// ─── Service ────────────────────────────────────────────────────────────────────

export class ActivityWatchService {
  private client: ActivityWatchClient;

  constructor(baseUrl: string) {
    this.client = new ActivityWatchClient(baseUrl);
  }

  async isRunning(): Promise<boolean> {
    return this.client.isRunning();
  }

  /**
   * Main entry point. Analyse a time window and return a result ready for storage.
   */
  async analyzeBucket(
    startTime: Date,
    endTime: Date,
    sessionWindow: ActivityBucket[] = []
  ): Promise<AnalysisResult | null> {
    try {
      const sortedWindow = [...sessionWindow].sort(
        (a, b) => new Date(a.startTime).getTime() - new Date(b.startTime).getTime()
      );

      const { category, dominantApp, apps, workSessionApp } =
        await this.analyzeWindowActivity(startTime, endTime, sortedWindow);

      return {
        startTime: startTime.toISOString(),
        endTime: endTime.toISOString(),
        category,
        dominantApp: category !== 'idle' ? dominantApp : null,
        apps,
        workSessionApp: workSessionApp || null,
      };
    } catch (error) {
      console.error('[AW] analyzeBucket error:', error);
      return null;
    }
  }

  // ─── Core analysis pipeline ─────────────────────────────────────────────────

  private async analyzeWindowActivity(
    startTime: Date,
    endTime: Date,
    sessionWindow: ActivityBucket[]
  ): Promise<{
    category: ActivityCategory;
    dominantApp: string;
    apps: Record<string, number>;
    workSessionApp: string;
  }> {
    const idle: { category: ActivityCategory; dominantApp: string; apps: Record<string, number>; workSessionApp: string } =
      { category: 'idle', dominantApp: '', apps: {}, workSessionApp: '' };

    try {
      console.log(`[AW] Analyzing ${startTime.toISOString()} → ${endTime.toISOString()}`);

      const events = await this.client.getCanonicalEvents(startTime, endTime);
      if (!events.length) return idle;

      const windowSeconds = (endTime.getTime() - startTime.getTime()) / 1000;
      const appUsage = this.calculateAppUsage(events);

      // Detect work session app first (even during idle, session context persists)
      const workSessionApp = this.detectWorkSessionApp(sessionWindow);

      if (this.isIdlePeriod(appUsage.apps, appUsage.totalActivity, windowSeconds)) {
        return { ...idle, apps: appUsage.apps, workSessionApp };
      }

      const patterns = this.analyzeAppPatterns(appUsage.apps, appUsage.totalActivity);
      const category = this.classifyActivityState(patterns, sessionWindow, workSessionApp);

      return { category, dominantApp: patterns.dominantApp, apps: appUsage.apps, workSessionApp };
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      console.error('[AW] analyzeWindowActivity error:', msg);
      return idle;
    }
  }

  // ─── App usage ──────────────────────────────────────────────────────────────

  private calculateAppUsage(events: AWEvent[]): { apps: Record<string, number>; totalActivity: number } {
    const apps: Record<string, number> = {};
    let totalActivity = 0;

    for (const event of events) {
      const app = (event.data?.app ?? '').toLowerCase();
      const dur = event.duration ?? 0;
      totalActivity += dur;
      if (app) apps[app] = (apps[app] ?? 0) + dur;
    }

    return { apps, totalActivity };
  }

  private isIdlePeriod(apps: Record<string, number>, totalActivity: number, windowSeconds: number): boolean {
    if (Object.keys(apps).length === 0 || windowSeconds <= 0) return true;
    return totalActivity / windowSeconds < CONFIG.IDLE_THRESHOLD_PERCENTAGE;
  }

  // ─── Pattern analysis ───────────────────────────────────────────────────────

  private analyzeAppPatterns(apps: Record<string, number>, totalActivity: number): AppUsagePattern {
    const empty: AppUsagePattern = {
      dominantApp: '', dominanceRatio: 0, appCount: 0, isProductive: false, sortedApps: [],
    };
    if (!Object.keys(apps).length || totalActivity <= 0) return empty;

    const sortedApps: [string, number][] = Object.entries(apps).sort((a, b) => b[1] - a[1]);
    const [dominantApp, dominantTime] = sortedApps[0];
    const dominanceRatio = dominantTime / totalActivity;

    return {
      dominantApp,
      dominanceRatio,
      appCount: sortedApps.length,
      isProductive: dominanceRatio >= CONFIG.PRODUCTIVE_SWITCHING_THRESHOLD,
      sortedApps,
    };
  }

  // ─── State classification ───────────────────────────────────────────────────

  private classifyActivityState(
    patterns: AppUsagePattern,
    sessionWindow: ActivityBucket[],
    workSessionApp: string
  ): ActivityCategory {
    const hasSwitching = this.detectAppSwitchingDistraction(sessionWindow, workSessionApp, patterns.sortedApps);

    // No history → first bucket
    if (!sessionWindow.length) {
      if (patterns.isProductive && !hasSwitching) return 'prefocus';
      if (this.isNeutral(patterns, hasSwitching)) return 'prefocus';
      return 'distraction';
    }

    // With history
    if ((patterns.isProductive || this.isNeutral(patterns, hasSwitching)) && !hasSwitching) {
      return this.handleProductiveState(this.extractRecentStates(sessionWindow));
    }

    return 'distraction';
  }

  /** Neutral = not dominant but few apps and no switching */
  private isNeutral(patterns: AppUsagePattern, hasSwitching: boolean): boolean {
    return !patterns.isProductive && !hasSwitching && patterns.appCount <= CONFIG.NEUTRAL_MAX_APP_COUNT;
  }

  private handleProductiveState(states: SessionWindowStates): ActivityCategory {
    // Maintain focus momentum
    if (states.last === 'focus') return 'focus';

    // Accumulated enough prefocus
    if (states.consecutivePrefocus >= CONFIG.PREFOCUS_BUCKETS_REQUIRED - 1) return 'focus';

    // Recover focus after brief distractions
    if (
      states.consecutiveUnproductive > 0 &&
      states.consecutiveUnproductive <= CONFIG.MAX_DISTRACTION_BUCKETS &&
      this.hadFocusBeforeDistractions(states.sequence)
    ) {
      return 'focus';
    }

    return 'prefocus';
  }

  // ─── Session window helpers ─────────────────────────────────────────────────

  private extractRecentStates(sessionWindow: ActivityBucket[]): SessionWindowStates {
    const sequence = sessionWindow.map((b) => b.category);
    const last = sequence.at(-1) ?? '';

    return {
      sequence,
      last,
      consecutivePrefocus: this.countConsecutiveFromEnd(sequence, ['prefocus']),
      consecutiveUnproductive: this.countConsecutiveFromEnd(sequence, ['distraction', 'idle']),
    };
  }

  private detectAppSwitchingDistraction(
    sessionWindow: ActivityBucket[],
    workSessionApp: string,
    currentSortedApps: [string, number][]
  ): boolean {
    if (sessionWindow.length < 2) return false;

    // Fingerprint = sorted top-N app names as string, e.g. "terminal,vscode"
    const toFingerprint = (apps: Record<string, number>): string =>
      Object.entries(apps)
        .sort((a, b) => b[1] - a[1])
        .slice(0, CONFIG.SWITCHING_TOP_N_APPS)
        .map(([name]) => name)
        .sort()
        .join(',');

    const fingerprints: string[] = [];

    // Current bucket (not yet stored)
    if (currentSortedApps.length > 0) {
      const currentApps = Object.fromEntries(currentSortedApps);
      const fp = toFingerprint(currentApps);
      if (fp) fingerprints.push(fp);
    }

    // Recent focused buckets from the same work session
    for (let i = sessionWindow.length - 1; i >= 0; i--) {
      const b = sessionWindow[i];
      if (!['focus', 'prefocus'].includes(b.category)) continue;
      if (b.workSessionApp && b.workSessionApp !== workSessionApp) break; // different session

      const fp = toFingerprint(this.parseAppsField(b.apps));
      if (fp) fingerprints.push(fp);
      if (fingerprints.length >= CONFIG.APP_SWITCHING_WINDOW_SIZE + 1) break;
    }

    // Need at least 2 to compare
    if (fingerprints.length < 2) return false;

    // High unique ratio = many different app combos = context switching
    const uniqueRatio = new Set(fingerprints).size / fingerprints.length;
    return uniqueRatio >= CONFIG.APP_SWITCHING_THRESHOLD;
  }

  /**
   * Detect the primary app being used across the current work session.
   * Only considers focus/prefocus buckets so distraction apps don't pollute.
   */
  private detectWorkSessionApp(sessionWindow: ActivityBucket[]): string {
    if (!sessionWindow.length) return '';

    const appTime: Record<string, number> = {};
    const focusBuckets = sessionWindow.filter(b => ['focus', 'prefocus'].includes(b.category));
    const bucketCount = focusBuckets.length;

    for (let idx = 0; idx < focusBuckets.length; idx++) {
      const bucket = focusBuckets[idx];

      const startMs = new Date(bucket.startTime).getTime();
      const endMs = new Date(bucket.endTime).getTime();
      const durationSec = (endMs - startMs) / 1000;
      if (durationSec <= 0) continue;

      // Recency weight: most recent bucket = 1.0, oldest = 0.4, linear decay
      const recencyWeight = bucketCount > 1
        ? 0.4 + 0.6 * (idx / (bucketCount - 1))
        : 1.0;

      const apps = this.parseAppsField(bucket.apps);
      for (const [appName, usageTime] of Object.entries(apps)) {
        if (usageTime / durationSec >= CONFIG.WORK_SESSION_APP_THRESHOLD_PERCENTAGE) {
          appTime[appName] = (appTime[appName] ?? 0) + usageTime * recencyWeight;
        }
      }
    }

    if (!Object.keys(appTime).length) return '';
    return Object.entries(appTime).sort((a, b) => b[1] - a[1])[0][0];
  }

  // ─── Utility ────────────────────────────────────────────────────────────────

  private countConsecutiveFromEnd(states: string[], targets: string[]): number {
    let count = 0;
    for (let i = states.length - 1; i >= 0; i--) {
      if (targets.includes(states[i])) count++;
      else break;
    }
    return count;
  }

  private hadFocusBeforeDistractions(states: string[]): boolean {
    if (!states.length) return false;

    let checkIdx = states.length;
    for (let i = states.length - 1; i >= 0; i--) {
      if (states[i] === 'distraction' || states[i] === 'idle') checkIdx = i;
      else break;
    }

    return checkIdx > 0 && states[checkIdx - 1] === 'focus';
  }

  /**
   * Type-safe parser for the `apps` JSON column.
   * The DB stores it as text (JSON), Drizzle normally returns a parsed object,
   * but guards against double-encoded strings and stringified numeric values.
   */
  private parseAppsField(apps: unknown): Record<string, number> {
    if (!apps) return {};

    // Handle double-encoded JSON string (e.g. the DB text wasn't auto-parsed)
    if (typeof apps === 'string') {
      try {
        apps = JSON.parse(apps);
      } catch {
        return {};
      }
    }

    if (typeof apps !== 'object' || Array.isArray(apps)) return {};

    const result: Record<string, number> = {};
    for (const [key, val] of Object.entries(apps as Record<string, unknown>)) {
      if (typeof val === 'number') {
        result[key] = val;
      } else if (typeof val === 'string') {
        const num = parseFloat(val);
        if (!Number.isNaN(num)) result[key] = num;
      }
    }
    return result;
  }
}
